File: Packages/TabCore/Sources/TabCore/PaymentCalculator.swift
import Foundation

public enum PaymentCalculatorError: Error, Equatable, Sendable {
    case emptyPayers
    case duplicatePayer(UUID)
    case unsupportedPaymentMode(PaymentMode)
    case exactAmountsRequired
    case missingAmountForPayer(UUID)
    case extraAmountForNonPayer(UUID)
    case amountHasTooManyFractionDigits(currency: String, maximumFractionDigits: Int)
    case amountsDoNotSumToTotal(expected: Decimal, actual: Decimal)
}

public enum PaymentCalculator {
    public static func calculate(
        totalAmount: Decimal,
        currency: String,
        payers: [UUID],
        paymentMode: PaymentMode,
        exactAmounts: [UUID: Decimal]? = nil
    ) throws -> [Payment] {
        guard !payers.isEmpty else {
            throw PaymentCalculatorError.emptyPayers
        }

        switch paymentMode {
        case .equal:
            let splits = try delegate(
                totalAmount: totalAmount,
                currency: currency,
                participants: payers,
                splitType: .equal,
                exactAmounts: nil
            )
            return splits.map {
                Payment(payerID: $0.participantID, amountPaid: $0.amountOwed, paymentMode: .equal)
            }

        case .exact:
            guard let amounts = exactAmounts else {
                throw PaymentCalculatorError.exactAmountsRequired
            }
            let splits = try delegate(
                totalAmount: totalAmount,
                currency: currency,
                participants: payers,
                splitType: .exact,
                exactAmounts: amounts
            )
            return splits.map {
                Payment(payerID: $0.participantID, amountPaid: $0.amountOwed, paymentMode: .exact)
            }

        case .percentage, .shares, .adjustment:
            throw PaymentCalculatorError.unsupportedPaymentMode(paymentMode)
        }
    }

    private static func delegate(
        totalAmount: Decimal,
        currency: String,
        participants: [UUID],
        splitType: SplitType,
        exactAmounts: [UUID: Decimal]?
    ) throws -> [ExpenseSplit] {
        do {
            return try SplitCalculator.calculate(
                totalAmount: totalAmount,
                currency: currency,
                participants: participants,
                splitType: splitType,
                exactAmounts: exactAmounts
            )
        } catch let err as SplitCalculatorError {
            throw map(err)
        }
    }

    private static func map(_ err: SplitCalculatorError) -> PaymentCalculatorError {
        switch err {
        case .emptyParticipants:
            return .emptyPayers
        case .duplicateParticipant(let id):
            return .duplicatePayer(id)
        case .unsupportedSplitType(let type):
            return .unsupportedPaymentMode(type)
        case .exactAmountsRequired:
            return .exactAmountsRequired
        case .missingAmountForParticipant(let id):
            return .missingAmountForPayer(id)
        case .extraAmountForNonParticipant(let id):
            return .extraAmountForNonPayer(id)
        case .amountHasTooManyFractionDigits(let currency, let maximumFractionDigits):
            return .amountHasTooManyFractionDigits(
                currency: currency,
                maximumFractionDigits: maximumFractionDigits
            )
        case .amountsDoNotSumToTotal(let expected, let actual):
            return .amountsDoNotSumToTotal(expected: expected, actual: actual)
        case .sharesRequired, .missingShareForParticipant, .extraShareForNonParticipant, .nonPositiveShare:
            // Unreachable: payments only delegate equal/exact splits.
            return .unsupportedPaymentMode(.shares)
        case .percentagesRequired, .missingPercentageForParticipant, .extraPercentageForNonParticipant,
             .nonPositivePercentage, .percentagesDoNotSumTo100:
            // Unreachable: payments only delegate equal/exact splits.
            return .unsupportedPaymentMode(.percentage)
        }
    }
}

File: Apps/TabAndroid/core/domain/src/main/kotlin/com/rithwikshetty/tab/domain/PaymentCalculator.kt
package com.rithwikshetty.tab.domain

import java.math.BigDecimal
import java.util.UUID

public sealed class PaymentCalculatorException(message: String) : IllegalArgumentException(message) {
    public data object EmptyPayers : PaymentCalculatorException("Payers cannot be empty.")
    public data class DuplicatePayer(public val id: UUID) :
        PaymentCalculatorException("Duplicate payer: $id")
    public data class UnsupportedPaymentMode(public val mode: SplitType) :
        PaymentCalculatorException("Unsupported payment mode: $mode")
    public data object ExactAmountsRequired : PaymentCalculatorException("Exact amounts are required.")
    public data class MissingAmount(public val id: UUID) :
        PaymentCalculatorException("Missing amount for payer: $id")
    public data class ExtraAmount(public val id: UUID) :
        PaymentCalculatorException("Amount supplied for non-payer: $id")
    public data class InvalidPrecision(public val currency: String, public val maximumFractionDigits: Int) :
        PaymentCalculatorException("Amount exceeds $currency precision.")
    public data class AmountsDoNotSum(public val expected: BigDecimal, public val actual: BigDecimal) :
        PaymentCalculatorException("Amounts do not sum to total.")
}

public object PaymentCalculator {
    public fun calculate(
        totalAmount: BigDecimal,
        currency: String,
        payers: List<UUID>,
        paymentMode: SplitType,
        exactAmounts: Map<UUID, BigDecimal>? = null,
    ): List<Payment> {
        if (paymentMode !in setOf(SplitType.EQUAL, SplitType.EXACT)) {
            throw PaymentCalculatorException.UnsupportedPaymentMode(paymentMode)
        }
        return try {
            SplitCalculator.calculate(
                totalAmount = totalAmount,
                currency = currency,
                participants = payers,
                splitType = paymentMode,
                exactAmounts = exactAmounts,
            ).map { Payment(it.participantId, it.amountOwed, paymentMode) }
        } catch (error: SplitCalculatorException) {
            throw map(error)
        }
    }

    private fun map(error: SplitCalculatorException): PaymentCalculatorException = when (error) {
        SplitCalculatorException.EmptyParticipants -> PaymentCalculatorException.EmptyPayers
        is SplitCalculatorException.DuplicateParticipant -> PaymentCalculatorException.DuplicatePayer(error.id)
        is SplitCalculatorException.UnsupportedSplitType ->
            PaymentCalculatorException.UnsupportedPaymentMode(error.type)
        SplitCalculatorException.ExactAmountsRequired -> PaymentCalculatorException.ExactAmountsRequired
        is SplitCalculatorException.MissingAmount -> PaymentCalculatorException.MissingAmount(error.id)
        is SplitCalculatorException.ExtraAmount -> PaymentCalculatorException.ExtraAmount(error.id)
        is SplitCalculatorException.InvalidPrecision ->
            PaymentCalculatorException.InvalidPrecision(error.currency, error.maximumFractionDigits)
        is SplitCalculatorException.AmountsDoNotSum ->
            PaymentCalculatorException.AmountsDoNotSum(error.expected, error.actual)
        SplitCalculatorException.SharesRequired,
        is SplitCalculatorException.MissingShare,
        is SplitCalculatorException.ExtraShare,
        is SplitCalculatorException.NonPositiveShare,
        -> PaymentCalculatorException.UnsupportedPaymentMode(SplitType.SHARES)
        SplitCalculatorException.PercentagesRequired,
        is SplitCalculatorException.MissingPercentage,
        is SplitCalculatorException.ExtraPercentage,
        is SplitCalculatorException.NonPositivePercentage,
        is SplitCalculatorException.PercentagesDoNotSumTo100,
        -> PaymentCalculatorException.UnsupportedPaymentMode(SplitType.PERCENTAGE)
    }
}
